File: lib/src/main/java/com/sd/lib/cache/simple/SimpleObjectCache.java
package com.sd.lib.cache.simple;

import com.sd.lib.cache.Cache;
import com.sd.lib.cache.CacheInfo;
import com.sd.lib.cache.handler.ObjectHandler;

public class SimpleObjectCache implements Cache.ObjectCache
{
    private final ObjectHandler mObjectHandler;

    public SimpleObjectCache(CacheInfo info)
    {
        mObjectHandler = new ObjectHandler(info)
        {
            @Override
            protected String getKeyPrefix()
            {
                return "object_";
            }
        };
    }

    @Override
    public boolean put(Object value)
    {
        final String key = value.getClass().getName();
        return mObjectHandler.putCache(key, value);
    }

    @Override
    public boolean put(Object value, Class<?> clazz)
    {
        if (clazz == null)
            return put(value);

        if (clazz.isInterface())
            throw new IllegalArgumentException("clazz is interface " + clazz);

        final Class<?> valueClass = value.getClass();
        if (!clazz.isAssignableFrom(valueClass))
            throw new IllegalArgumentException(clazz + " is not assignable from " + valueClass);

        final String key = clazz.getName();
        return mObjectHandler.putCache(key, value);
    }

    @Override
    public <T> T get(Class<T> clazz)
    {
        final String key = clazz.getName();
        return (T) mObjectHandler.getCache(key, clazz);
    }

    @Override
    public boolean remove(Class<?> clazz)
    {
        final String key = clazz.getName();
        return mObjectHandler.removeCache(key);
    }

    @Override
    public boolean contains(Class<?> clazz)
    {
        final String key = clazz.getName();
        return mObjectHandler.containsCache(key);
    }
}

File: lib/src/main/java/com/sd/lib/cache/simple/SimpleObjectCache.kt
package com.sd.lib.cache.simple

import com.sd.lib.cache.Cache.ObjectCache
import com.sd.lib.cache.CacheInfo
import com.sd.lib.cache.handler.impl.ObjectHandler

internal class SimpleObjectCache(info: CacheInfo) : ObjectCache {
    private val _handler = ObjectHandler(info, "object")

    override fun put(value: Any?): Boolean {
        if (value == null) return false
        val key = value.javaClass.name
        return _handler.putCache(key, value)
    }

    @Suppress("UNCHECKED_CAST")
    override fun <T> get(clazz: Class<T>): T? {
        val key = clazz.name
        val cache = _handler.getCache(key, clazz) ?: return null
        return cache as T
    }

    override fun remove(clazz: Class<*>): Boolean {
        val key = clazz.name
        return _handler.removeCache(key)
    }

    override fun contains(clazz: Class<*>): Boolean {
        val key = clazz.name
        return _handler.containsCache(key)
    }
}